yapapi/executor/__init__.py

```diff
@@ -557,22 +566,66 @@ async def start_worker(agreement: rest.market.Agreement, node_info: NodeInfo) ->
                             )
                             cmds = cc.commands()
                             emit(events.ScriptSent(agr_id=agreement.id, task_id=task_id, cmds=cmds))
 
-                            async for evt_ctx in remote:
-                                evt = evt_ctx.event(agr_id=agreement.id, task_id=task_id, cmds=cmds)
-                                emit(evt)
-                                if isinstance(evt, events.CommandExecuted) and not evt.success:
-                                    raise CommandExecutionError(evt.command, evt.message)
+                            async def get_batch_results():
+
+                                print(colors.yellow("[get_batch_results] Waiting for results..."))
 
-                            emit(events.GettingResults(agr_id=agreement.id, task_id=task_id))
-                            await batch.post()
-                            emit(events.ScriptFinished(agr_id=agreement.id, task_id=task_id))
-                            await accept_payment_for_agreement(agreement.id, partial=True)
+                                results = []
+                                async for evt_ctx in remote:
+                                    evt = evt_ctx.event(
+                                        agr_id=agreement.id, task_id=task_id, cmds=cmds
+                                    )
+                                    emit(evt)
+                                    results.append(evt)
+                                    if isinstance(evt, events.CommandExecuted) and not evt.success:
+                                        raise CommandExecutionError(evt.command, evt.message)
+
+                                emit(events.GettingResults(agr_id=agreement.id, task_id=task_id))
+                                await batch.post()
+                                emit(events.ScriptFinished(agr_id=agreement.id, task_id=task_id))
+                                await accept_payment_for_agreement(agreement.id, partial=True)
+
+                                print(colors.yellow("[get_batch_results] Got results"))
+                                return results
+
+                            if batch.wait_for_results or batch.contains_init_step:
+                                # Block until the results are available
+                                print(
+                                    colors.yellow(
+                                        "[executor] Blocking until results are available..."
+                                    )
+                                )
+                                results = await get_batch_results()
+                                print(colors.yellow("[executor] Returning results"))
+
+                                async def make_awaitable():
+                                    return results
+
+                                batch = await command_generator.asend(make_awaitable())
+                            else:
```

Comment:
plus:

```suggestion
                            else:
                                assert not batch.contains_init_step, "init step must be blocking"
```

Reply:
Hmm, this looks like we're trying to `assert` some properties of data that we do not control, namely the `batch` argument. 
I think raising an exception would be more appropriate if this condition is not satisfied.

Another approach I think would be to leave the code as is (i.e. override the `blocking` flag if the batch contains initialization commands) and document the `blocking` flag saying something like: 
```
`future_results = yield ctx.commit(blocking=False)` returns without waiting
for batch results _whenever possible_, if the batch contains `Deploy` or `Start`
commands, `yield` will block until they are executed.
```

Reply:
okay, from the discussion with @azawlocki :

* it would be the user's responsibility to know whether `deploy` and `start` should be blocking and if they choose to run them in a non-blocking way, it's on them to cope with the exception (or lack of it -> because some other exeunit may not require it)
* plus, we'd add it to the documentation of the VM payload that vm exeunit requires the deploy/start to finish before one can issue any other commands